plugins/destination/postgresql/client/insert.go

```diff
@@ -34,8 +34,12 @@ func (c *Client) InsertBatch(ctx context.Context, messages message.WriteInserts)
 		return err
 	}
 
-	var sql string
 	batch := &pgx.Batch{}
+
+	// Queries cache.
+	// We may consider LRU cache in the future, but even for 10K records it may be OK to just save.
+	queries := make(map[string]string, 100)
```

Comment:
Can we guarantee that every insert record has the same query? This would stop users from using different versions of the same plugin in the same configuration.

if a user is using `deterministic_cq_id: true` and `pk_mode: cq_id_only` then they can safely ignore PK changes and therefore  could mix plugin versions

Reply:
If this is an optimization we really want to use we could use all of the cq_meta data fields as cache keys rather than just the table name

Reply:
We can guarantee, as the migrate will first flush the inserts.

Reply:
Not if they run with the flag `--no-migrate`

Reply:
In this case we still refresh the info with list tables call at the beginning of the flush, so we're also OK there

Reply:
Why would we be doing a flush if we are not doing a migrate?

Reply:
PG plugin uses ([mixed](https://github.com/cloudquery/cloudquery/blob/main/plugins/destination/postgresql/client/client.go#L95)) batch writer, so we always perform writing in a batch (to different tables in a single batch, but still).

When performing flush (writing a batch) we first check if we need to [refresh info about tables](https://github.com/cloudquery/cloudquery/blob/main/plugins/destination/postgresql/client/insert.go#L24-L30). This data is either filled in during migration phase to normalize tables (& updated when we create new table), or updated right before insert.

So, in both simple & `--no-migrate` modes we will end up with already properly filled in info for normalized tables.
Additionally, when [looking up the table to insert to](https://github.com/cloudquery/cloudquery/blob/main/plugins/destination/postgresql/client/insert.go#L49) we use `schema.Tables.Get` that will just perform a lookup in a list & return the first encountered entry, so the schema returned is always the same.

Reply:
This config I don't think we work as a single batch can have records from different tables.
```
kind: source
spec:
  # Source spec section
  name: aws-1
  version: version: v21.0.1
  deterministic_cq_id: true
  tables:
    - aws_ec2_images
  destinations: ["postgresql"]
  spec:
    org:
      member_role_name: OrganizationAccountAccessRole
---
kind: source
spec:
  name: aws-2
  version: v22.0.1
  deterministic_cq_id: true
  tables:
    # This table added new columns in v22.0.1
    - aws_ec2_images
  destinations: ["postgresql"]
  spec:
    accounts:
      - id: test
        local_profile: test-profile
---
kind: destination
spec:
  name: "postgresql"
  path: "cloudquery/postgresql"
  registry: "github"
  version: "v6.1.3"
  pk_mode: cq-id-only
  migrate_mode: forced
  spec:
    connection_string: "postgresql://postgres:pass@localhost:5432/postgres?sslmode=disable"
```

Notice how in the 2 source configs they are syncing the same table name, but because they are from different versions so there could be columns included in `aws-2` that are not in `aws-1`... This is definitely an edge case but it is a valid case that we have to consider...

Reply:
@bbernays we don't sync them in parallel. (also see sync_v3.go & flushes between those parts)